### PR TITLE
Indirectly DKIM-protect Autocrypt key material with Autocrypt-Auth headers

### PR DESCRIPTION
**Pull request by Valodim**
Following up the discussion in #387, this PR is a concrete proposal for DKIM-authenticating Autocrypt key material.

* To avoid the need for a complicated canonicalization method, only the key material is authenticated

* This uses a full sha2, adding a constant 60 bytes of overhead to each message (per Autocrypt header). I would personally consider a half sha2 sufficient since this is a preimage attack scenario, but people are weary of truncating hashes so I went with the conservative option here.

* Encrypted messages are also signed, so the auth tag would be redundant there.

### CODE REVIEW

File: doc/level1.rst [hunk 1]
```diff
@@ -723,6 +723,47 @@ way:
...
++++++++++++++++++++
+
+To combat spam, phishing, and other types of spoofed messages, many
+providers authenticate all outgoing messages by adding a DomainKeys
+Identified Mail (:rfc:`DKIM<6376>`) signature.  This signature is
+created over the body of the message, and a configurable set of its
+headers. This set typically includes headers like ``From``, ``To``,
+``Date`` and ``Subject``, but not ``Autocrypt``.
+
+Providers can't be expected to include Autocrypt in the set of signed
```
[the remark this answers is not in the document]
Comment (Valodim) [reply doc/level1.rst]: Hmh. Not a fan of giving shout-outs in the spec like that, personally

**Comment by Valodim** [comment]
> I think this idea is interesting, but i'm still missing the bigger picture -- what is a client supposed to do when something like this is DKIM-signed, or if it is not DKIM-signed? what kinds of verification are supposed to happen, and what are the consequences of that verification passing or failing?

In order to tamper with the Autocrypt headers (or at least their public keys) of a message, an attacker will have to either:
1) modify only the Autocrypt header, introducing an inconsistency with the Autocrypt-Auth tag. In this case, the Autocrypt header will be rejected by the MUA (see third last paragraph)
2) modify both the Autocrypt and Autocrypt-Auth headers, which will invalidate the DKIM signature, causing the receiving provider to treat the message as suspicious (see second last paragraph). What this means in practice varies: with strict DMARC policies (Yahoo, AOL) the message would be straight rejected, and I think GMail displays a warning to the user, but hopefully in all cases there is a high chance it'll land in the junk folder

> who is this intended to defend against?

It's intended as a direct answer to the question "But can't anyone just insert headers then?", which does come up every now and then in my experience of talking to people about Autocrypt. The benefits should be almost the same as if the Autocrypt header was DKIM-signed, without relying on provider support

**Comment by dkg** [comment]
thanks for doing this round of cleanup, @Valodim !  i think my main questions are still unanswered, though.

In particular:

 * Are we suggesting that clients do DKIM validation themselves?
 * If clients do not do DKIM validation, then who is this defending against?  
    * If it is defending against active manipulation by the sending MTA, then it doesn't work, because the sender's MTA is the one doing the DKIM signing -- they can just modify the appropriate fields before signing.
    * if it is defending against the receiving MTA, then it doesn't work, because the recipient's MTA can just modify the message after receipt, without caring that the change breaks the DKIM header, since the client is not validating.
    * is it defending against some attacker *other* than the sending or receiving MTA?  if so, who?
 * On the other hand, if we expect the client to do DKIM validation, are we expecting that they should ignore Autocrypt headers if the DKIM validation fails?  What should a DKIM-validating client do if no DKIM signature exists on the message?

**Comment by Valodim** [comment]
> Are we suggesting that clients do DKIM validation themselves?

Definitely not, DKIM validation is not something that clients can reasonably do. The increase in security comes from 1) rejection of autocrypt headers if auth tag doesn't match + 2) increased spam score if dkim signature is stripped + 3) increased risk of exposure for provider if manipulative dkim signature is created

The gain in security is similar to how the trust is tied to the provider and X509 PKI when using WKD. Of course it's not fully secure, but tying the trust to DKIM is a lot better than nothing.

Honestly, this might introduce more complexity than it's worth for the benefit it adds. But the reason I think this might be worth it is that I've often heard the concern that "anyone" could just inject Autocrypt headers, I guess because people have been taught by e2e campaigns that emails in transit can generally be very easily manipulated. If we're able to answer "Autocrypt headers are tied to the providers' DKIM signature", we have a much better answer to that concern.

**Comment by dkg** [comment]
@azul wrote:
> if the receiving MTA checks them

i think it's "checks them *and* refuses messages without DKIM signatures"

so we're talking about the sending domain needing a strict DMARC policy as well, right?

**Comment by Valodim** [comment]
This is an interesting idea still, but it seems like the benefits aren't concrete enough to warrant the complexity. I'll shelve this for now.

**Pull request closed without merging**